crates/sui-core/src/quorum_driver/mod.rs

```diff
@@ -221,6 +274,163 @@ where
+    async fn attempt_one_conflicting_transaction(
+        &self,
+        tx_digest: &&TransactionDigest,
+        original_tx_digest: &TransactionDigest,
+        validators: Vec<&AuthorityName>,
+    ) -> SuiResult<bool> {
+        let clients = self.validators.load();
+        for validator_name in validators {
+            // If we cannot find the client, it indicates an epoch change. Then we stop all attempts.
+            let client = clients.get_client(validator_name).ok_or_else( || {
+                info!(?tx_digest, "It looks like we have an epoch change when doing attempt_one_conflicting_transaction.");
+                SuiError::InconsistentEpochState { error: format!("Epoch advance caused validator {:?} missing in AuthorityAggreagtor, giving up all attempts.", validator_name) }
+            })?;
+            if let Ok(TransactionInfoResponse {
+                signed_transaction,
+                certified_transaction,
+                signed_effects: _,
+            }) = client
+                .handle_transaction_info_request(TransactionInfoRequest {
+                    transaction_digest: **tx_digest,
+                })
+                .await
+            {
+                // If we happen to find that a validator returns TransactionCertificate:
+                if let Some(certified_transaction) = certified_transaction {
+                    self.metrics
+                        .total_times_conflicting_transaction_already_finalized_when_retrying
+                        .inc();
+                    // We still want to ask validators to execute this certificate in case this certificate is not
+                    // known to the rest of them (e.g. when *this* validator is bad).
+                    let result = self
+                        .validators
+                        .load()
+                        .process_certificate(certified_transaction.into_inner())
+                        .await
+                        .tap_ok(|_resp| {
+                            debug!(
+                                ?tx_digest,
+                                ?original_tx_digest,
+                                "Retry conflicting transaction certificate succeeded."
+                            );
+                        })
+                        .tap_err(|err| {
+                            debug!(
+                                ?tx_digest,
+                                ?original_tx_digest,
+                                "Retry conflicting transaction certificate got an error: {:?}",
+                                err
+                            );
+                        });
+                    // We only retry once.
+                    return Ok(result.is_ok());
+                }
+                if let Some(verified_transaction) = signed_transaction {
+                    let transaction =
+                        TransactionEnvelope::from(verified_transaction).remove_auth_sig_info();
```

Comment:
I think this can be done with `into_inner().to_transaction()`?

Reply:
you are right